### PR TITLE
Adding custom algorithm

### PR DESCRIPTION
**Pull request by Teukros**
This pull request introduce possibility to use your own algorithm with jsonwebtoken library. If you pass proper options you can use sign, verify and decode methods together with chosen algorithm function.

### CODE REVIEW

File: sign.js [hunk 1]
```diff
@@ -13,14 +14,15 @@ var sign_options_schema = {
...
+  algorithm: { isValid: includes.bind(null, ['RS256', 'RS384', 'RS512', 'ES256', 'ES384', 'ES512', 'HS256', 'HS384', 'HS512', 'custom', 'none']), message: '"algorithm" must be a valid string enum value' },
   header: { isValid: isPlainObject, message: '"header" must be an object' },
   encoding: { isValid: isString, message: '"encoding" must be a string' },
   issuer: { isValid: isString, message: '"issuer" must be a string' },
   subject: { isValid: isString, message: '"subject" must be a string' },
   jwtid: { isValid: isString, message: '"jwtid" must be a string' },
   noTimestamp: { isValid: isBoolean, message: '"noTimestamp" must be a boolean' },
-  keyid: { isValid: isString, message: '"keyid" must be a string' }
+  keyid: { isValid: isString, message: '"keyid" must be a string' },
+  customAlgorithmFunction: {isValid: isFunction, message: "custom algorithm must provide it's own sign function"}
```
Comment (montgomery1944) [inline sign.js]: it's -> its
" -> '

File: sign.js [hunk 2]
```diff
@@ -169,20 +171,30 @@ module.exports = function (payload, secretOrPrivateKey, options, callback) {
...
-    jws.createSign({
-      header: header,
-      privateKey: secretOrPrivateKey,
-      payload: payload,
-      encoding: encoding
-    }).once('error', callback)
-      .once('done', function (signature) {
-        callback(null, signature);
+    if (isCustomAlgorithmUsed) {
+      options.customAlgorithmFunction(payload, secretOrPrivateKey, options, function(err, result) {
```
Comment (montgomery1944) [inline sign.js]: Formatting.

File: verify.js [hunk 3]
```diff
@@ -29,6 +30,15 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
     };
   }
 
+  if (options.algorithms && includes(options.algorithms, 'custom')) {
+    try {
+      decodedToken = options.customAlgorithmFunction(jwtString, secretOrPublicKey, options);
+      return done(null, decodedToken);
+    } catch(err) {
```
Comment (montgomery1944) [inline verify.js]: Formatting.

File: README.md [hunk 4]
```diff
@@ -203,6 +234,8 @@ __Warning:__ This will __not__ verify whether the signature is valid. You should
 
 * `json`: force JSON.parse on the payload even if the header doesn't contain `"typ":"JWT"`.
 * `complete`: return an object with the decoded payload and header.
+* `isCustomAlgorithmUsed`: boolean that declares if you want to use your own algorithm for decoding.
+* `customAlgorithmFunction`: function which will be used for decoding, if isCustomAlgorithmUsed is true
```
Comment (montgomery1944) [inline README.md]: Missing ".".

**Comment by rochdev** [comment]
I think it should be possible to use a custom signing function even when using a predefined algorithm.

My use case: https://github.com/auth0/node-jsonwebtoken/issues/427#issuecomment-356363954

**Comment by ziluvatar** [comment]
Sorry, I still have some concerns about this, what is the benefit of adding custom signing to this library? (at least how it is in the PR so far)

- On signing: I can see a little: some payload generation. Signer ends up even filling the header part. The function could even avoid adding any header or anything it could bypass the JWT format and build any other format.
- On verifying: It just get a string and return the object.

It seems to me like a simple serializer / deserializer. We should think how to offer this as simple as possible to the consumer and be sure that we are still in JWT playground.

If you take a look to `jws` package, what you want is to use your own `algo`: https://github.com/brianloveswords/node-jws/blob/master/lib/sign-stream.js#L21, so to generate the last part of the JWT (header.payload.**signature**). Is that right?

There is an open ticket there about your own use case: https://github.com/brianloveswords/node-jws/issues/34. I think it would have more sense to have it supported in there.

If we still wanted to support it here we would need to write, or bring, most of that to this library, having here two signers with same interface `custom` and `jws` and leaving the final custom signing function for the consumer similar to what they do in `jws`: `algo(securedInput, secretOrKey)` where `securedInput` is "header.payload" in base64 with encoding in mind.
If we don't do all of that, consumer would need to build `jws` basically to do it right.

What if you try to bump up the topic on jws repo and we pause this for now? Thoughts?

**Comment by novemberborn** [comment]
> what is the benefit of adding custom signing to this library?

I'm currently investigating custom algorithms. In my use case I have a secp256k1 private key I want to use to sign a JWT. The JWT would contain a claim of the corresponding public key.

Interface wise I think the `algorithms` option here should be an array that accepts strings and objects. For objects, keys would be the custom algorithm name and the value either a signer or verifier:

```js
jwt.sign(payload, secretOrPrivateKey, {
  algorithm: {
    'urn:foo:bar:custom:yay' (input) {
      return sign(input)
    }
  }
})

jwt.verify(token, publicKey, {
  algorithms: [
    {
      'urn:foo:bar:custom:yay' (input) {
        return verify(input)
      }
    }
  ]
})
```

This library could select the correct `algorithm` when verifying. I agree that it makes sense for `jws` to accept a custom function. Though that project doesn't look particularly active.

**Comment by ziluvatar** [comment]
There was a PR about it in the past: https://github.com/brianloveswords/node-jws/pull/35 I bet it was closed since some feedback was provided but no reaction was made by the PR creator.
Maybe someone can re-create it applying the feedback.

**Comment by panva** [comment]
JWT algorithms are well defined and we're committed to support the defined ones, as soon as the `jws` dependency supports them too.

`secp256k1` curve EC keys and alg name registration is only just being [standardized](https://tools.ietf.org/html/draft-ietf-cose-webauthn-algorithms-00) and is still very much in [flux](https://mailarchive.ietf.org/arch/msg/cose/91MouVA43DefbpQOB7l5daCEeIc)

The only other missing defined DSA JWA is EdDSA which is available in node 12 and will eventually make it to `jsonwebtoken` too.

Also as a general recommendation you should not roll your own crypto, which this very much looks like.

**Pull request closed without merging**